### ISSUE TITLE
Comparison: Diffusion-LM vs Current Text Diffusion Implementation

### ISSUE THREAD

**Issue by LiamConnell** [LiamConnell/TinyDiffusionModels#12]
# Diffusion-LM vs Current Text Diffusion Implementation

After analyzing the Diffusion-LM paper (arXiv:2205.14217) and comparing it with our current `src/shakespeare.py` implementation, I've identified several key architectural and methodological differences.

### Core Architecture
- **Non-autoregressive**: Generates entire sequences simultaneously
- **Trained rounding function**: Uses a learned softmax distribution to predict tokens from embeddings
- **Controllable generation**: Gradient-based control during the diffusion process
- **Fluency regularization**: Combines control objectives with fluency constraints: `λ log p(x_{t-1} < /dev/null | x) + log p(c|x_{t-1})`

### Technical Details
- Iteratively denoises Gaussian vectors into word vectors
- Maintains continuous latent variables throughout most of the process
- Uses minimum Bayesian risk decoding to select best samples
- Applies gradient updates on control objectives during generation

### Core Architecture
- **Pure diffusion sampling**: Standard DDPM approach without learned rounding
- **Cosine similarity decoding**: Simple nearest-neighbor approach using `F.normalize` and `torch.matmul`
- **Hybrid generation mode**: Optional AR+diffusion mixing via `guided_generate()`
- **Embedding space**: Operates in pre-trained model embedding space (Gemma-2b-it)

### Key Components
```python
# Current decoding approach
emb_norm = F.normalize(embed_matrix, dim=1)  # (V, dim)
x_norm = F.normalize(x, dim=2)               # (B, L, dim)
sims = torch.matmul(x_norm, emb_norm.T)      # (B, L, V)
tokens = sims.argmax(dim=-1)                 # (B, L)
```

## Key Differences

| Aspect | Diffusion-LM | Current Implementation |
|--------|--------------|----------------------|
| **Token Decoding** | Learned softmax rounding function | Cosine similarity + argmax |
| **Generation Mode** | Non-autoregressive only | Pure diffusion + optional AR hybrid |
| **Control** | Gradient-based during diffusion | Alpha-weighted logit mixing |
| **Training Target** | Custom embedding space | Pre-trained model embeddings |
| **Fluency** | Explicit regularization term | Implicit via pre-trained embeddings |

### 1. Learned Rounding Function
- Replace cosine similarity with trainable token prediction
- Could improve semantic coherence of generated text
- Would require additional training objective

### 2. Control Mechanisms
- Implement gradient-based control during sampling
- Add fluency regularization terms
- Enable fine-grained linguistic control (syntax, sentiment, etc.)

### 3. Architecture Changes
```python
class LearnedRounding(nn.Module):
    def __init__(self, embed_dim, vocab_size):
        super().__init__()
        self.decoder = nn.Linear(embed_dim, vocab_size)
    
    def forward(self, embeddings):
        return self.decoder(embeddings)  # (B, L, V)
```

## Implementation Questions

1. **Training Complexity**: Diffusion-LM requires training both the diffusion model and rounding function - worth the added complexity?

2. **Embedding Space**: Should we move away from pre-trained embeddings to a learned embedding space?

3. **Control vs Generation Quality**: Our current hybrid approach offers some control - how much improvement would full Diffusion-LM control provide?

4. **Computational Cost**: Gradient-based control during sampling may be expensive - acceptable for our use case?

## Next Steps

- [ ] Experiment with learned rounding function
- [ ] Implement gradient-based control sampling
- [ ] Compare generation quality: cosine similarity vs learned decoding
- [ ] Evaluate computational overhead of Diffusion-LM approach
- [ ] Test control capabilities on specific linguistic tasks

## Files Referenced
- `src/shakespeare.py:149-153` (current cosine similarity decoding)
- `src/shakespeare.py:179-207` (guided generation implementation)
- Diffusion-LM paper: https://arxiv.org/abs/2205.14217
